File: analysis/scripts/compute_public_goods_outcomes.R
args <- commandArgs(trailingOnly = TRUE)
if (length(args) < 2) {
  stop("Usage: Rscript compute_public_goods_outcomes.R input.csv output.csv [endowment] [multiplier]")
}

data <- read.csv(args[[1]], stringsAsFactors = FALSE)
endowment <- if (length(args) >= 3) as.numeric(args[[3]]) else 10
multiplier <- if (length(args) >= 4) as.numeric(args[[4]]) else 1.5
keys <- c("game_id", "group_id", "round")
split_rows <- split(data, interaction(data[keys], drop = TRUE))

rows <- lapply(split_rows, function(group) {
  contribution <- as.numeric(if ("action_value" %in% names(group)) group$action_value else group$contribution)
  total <- sum(contribution, na.rm = TRUE)
  public_return <- total * multiplier / nrow(group)
  group$contribution <- contribution
  group$total_contribution <- total
  group$average_contribution <- total / nrow(group)
  group$public_return <- public_return
  group$computed_payoff <- endowment - contribution + public_return
  group
})

output <- do.call(rbind, rows)
dir.create(dirname(args[[2]]), recursive = TRUE, showWarnings = FALSE)
write.csv(output, args[[2]], row.names = FALSE)

File: analysis/scripts/compute_public_goods_outcomes.py
#!/usr/bin/env python3
"""Compute public goods payoffs from player-round contribution data."""

from __future__ import annotations

import argparse

from _common import group_by, number, read_records, write_records


def compute(records: list[dict], endowment: float, multiplier: float) -> list[dict]:
    output = []
    for key, grouped in group_by(records, ["game_id", "group_id", "round"]).items():
        contributions = [number(row.get("action_value", row.get("contribution"))) for row in grouped]
        total = sum(contributions)
        public_return = total * multiplier / len(grouped) if grouped else 0
        for row, contribution in zip(grouped, contributions):
            payoff = endowment - contribution + public_return
            output.append(
                {
                    **row,
                    "contribution": contribution,
                    "total_contribution": total,
                    "average_contribution": total / len(grouped) if grouped else "",
                    "public_return": public_return,
                    "computed_payoff": payoff,
                    "game_id": key[0],
                    "group_id": key[1],
                    "round": key[2],
                }
            )
    return output


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("input")
    parser.add_argument("output")
    parser.add_argument("--endowment", type=float, default=10)
    parser.add_argument("--multiplier", type=float, default=1.5)
    args = parser.parse_args()
    write_records(args.output, compute(read_records(args.input), args.endowment, args.multiplier))


if __name__ == "__main__":
    main()

